src/Icicle/Source/Transform/Inline.hs

```diff
@@ -35,18 +41,27 @@ inlineTransform funs
   tranx _ x
    | (Var a n, args) <- takeApps x
    , Just fun        <- Map.lookup n funs
-   , length args == length (arguments fun)
-   = do let argNames
-                    = fmap snd $ arguments fun
-        ns         <- mapM (freshPrefixBase . nameBase) argNames
+   , vars            <- arguments fun
+   , argNames        <- fmap snd vars
+   , length args == length vars
+   = case opt of
+      InlineByName -> do
+       ns      <- mapM (freshPrefixBase . nameBase) argNames
+
+       let lets = fmap (mkLet a) (ns `zip` args)
+       let sub  = Map.fromList
+                $ argNames `zip` fmap (Var a) ns
 
-        let sub     = Map.fromList
-                    $ argNames `zip` fmap (Var a) ns
+       body'   <- substQ sub $ body fun
 
-        body'      <- substQ sub $ body fun
+       return ((), Nested a (prefixContexts lets body'))
 
-        let lets = fmap (mkLet a) (ns `zip` args)
-        return ((), Nested a (prefixContexts lets body'))
+      InlineByValue -> do
+       let sub  = Map.fromList
+                $ argNames `zip` args
+       body'   <- substQ sub $ body fun
+
+       return ((), Nested a body')
```

Comment:
yep, this looks good.
I found the names a bit confusing at first because I was thinking about call-by-name and call-by-value, but call-by-value is inline-by-name and call-by-name is inline-by-value. does that make sense? I can't think of any better names right now though